### PR TITLE
Support Python 3.12 on tests for core modules

### PR DESCRIPTION
**Pull request by not522**
## Motivation
Adding CI to support Python 3.12. #5000
Since BoTorch does not yet support Python 3.12, CI ignores it.

## Description of the changes
- Add Python 3.12 in `tests.yml`
- Add badge in `README.md`
- Update classifiers in `pyproject.toml`

### CODE REVIEW

**Review by HideakiImamura — COMMENTED** [review]
Let me clarify the scope of this PR. Is this PR work on the first TODO in https://github.com/optuna/optuna/issues/5000? Is so, I think we should modify the title and undo the change in the integration tests.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -77,13 +77,27 @@ jobs:
...
       run: |
-        pytest tests -m "not integration and not slow"
+        if [ ${{ matrix.python-version }} == 3.12 ]; then
+          # TODO(not522): Remove ignores when BoTorch supports Python 3.12
+          pytest tests -m "not integration and not slow" \
+            --ignore tests/terminator_tests/ \
+            --ignore tests/visualization_tests/test_terminator_improvement.py
+        else
+          pytest tests -m "not integration and not slow"
+        fi
```
Comment (nabenabe0928) [inline .github/workflows/tests.yml:101]: What about this?
I bundled "Tests" and "Scheduled tests", and then the maintenance cost will lower.

```yaml
    - name: Tests
      run: |
        if [ "${{ github.event_name }}" = "schedule" ] || \
           [ "${{ github.event_name }}" = "workflow_dispatch" ] ; then
          target="not integration"
        else
          target="not integration and not slow"
        fi

        if [ "${{ matrix.python-version }}" = "3.12" ] ; then
          # TODO(not522): Remove ignores when BoTorch supports Python 3.12
          ignore_botorch_option="--ignore tests/terminator_tests/ --ignore \
          tests/visualization_tests/test_terminator_improvement.py"
        else
          ignore_botorch_option=""
        fi

        pytest tests -m "$target" $ignore_botorch_option
```

**Comment by not522** [comment]
> Let me clarify the scope of this PR. Is this PR work on the first TODO in https://github.com/optuna/optuna/issues/5000? Is so, I think we should modify the title and undo the change in the integration tests.

Thank you for your review. I updated the title. This PR handles first and last two TODOs.
The change for `test_botorch.py` is necessary to avoid `ImportError`.

**Review by nabenabe0928 — APPROVED** [review]
LGTM:)

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Pull request merged**